### DISCUSSION TITLE
How to configure Renovate to use angular style commit messages for breaking changes

### DISCUSSION THREAD

**Discussion by skycaptain** [renovatebot/renovate#33021] [Request Help] [answered]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

Self-hosted GitLab

### Please tell us more about your question or problem

I have a project that uses [semantic-releases](https://github.com/semantic-release/semantic-release) and has a dependency `mylibrary` managed by Renovate. The project uses Angular-style commit messages, which are enforced using [commitlint](https://commitlint.js.org/).

For `mylibrary` updates, I want Renovate to create different semantic commit types: a `fix` type for patch updates, a `feat` type for minor updates, and a breaking change commit for major updates. My config works like this:

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": ["config:best-practices"],
  "packageRules": [
    {
      "description": "Trigger breaking release for major updates",
      "matchPackageNames": ["mylibrary"],
      "matchUpdateTypes": ["major"],
      "commitMessagePrefix": "feat(deps)!:"
    },
    {
      "description": "Trigger feature release for minor updates",
      "matchPackageNames": ["mylibrary"],
      "matchUpdateTypes": ["minor"],
      "semanticCommitType": "feat"
    },
    {
      "description": "Trigger fix release for patch updates",
      "matchPackageNames": ["mylibrary"],
      "matchUpdateTypes": ["patch"],
      "semanticCommitType": "fix"
    }
  ]
}
```

While this configuration mostly works, I've noticed an issue: for minor and patch updates, Renovate creates commit messages like `feat(deps): update dependency ...`, but for major updates it uses `feat(deps)!: Update dependency`, which fails commitlint validation (notice uppercase `U`). It appears that setting a `commitMessagePrefix` interferes with semantic commit detection. What would be the best solution here? Would enforcing `semanticCommits` in the package rule be sufficient? Is there a more elegant approach to implementing these package rules?

### Logs (if relevant)

<details><summary>Logs</summary>

```

Replace this text with your logs, between the starting and ending triple backticks

```

</details>

**Comment by rarkins** [comment]
I'm not sure it's possible to get it working as-is, but if not then the best solution would probably be to support configuration like this: `"semanticCommitType": "feat!"`

**Reply by skycaptain** [reply]
Thanks for having a look. Right now, setting `"semanticCommitType": "feat!"`  would render to `feat!(deps): update ...`, which has incorrect syntax (the `!` must be placed immediately before the `:` according to the spec). We could however additionally set `"semanticCommitScope": ""` to generate a valid string `feat!: update ...` loosing the scope however.

Another solution could be defining `"commitMessage": "{{{commitMessagePrefix}}} {{{lowercase commitMessageAction}}} {{{commitMessageTopic}}} {{{commitMessageExtra}}} {{{commitMessageSuffix}}}"`. This produces a valid commit message, though the behaviour differs slightly since `commitMessageTopic` and `commitMessageExtra` aren't lowercased. For example, it renders as `feat(deps)!: update myimage Docker tag to v5` rather than `feat(deps)!: update myimage docker tag to v5`. Currently, the `lowercase` helper seems to only work with `commitMessageAction` in this scenario. When applied to `commitMessageTopic` or `commitMessageExtra`, it produces an empty string—possibly because these fields are templates themselves?

A third option would be to allow the `commitMessageLowerCase` option to be permanently `enabled` to enforce lowercase formatting.

**Reply by bodgit** [reply]
I just bumped into this issue; I want to be able to force a particular dependency to always generate a breaking change. I figured I'd give setting `"semanticCommitType": "fix!"` a go and be pleasantly surprised if it generated a valid semantic commit but was fully expecting the actual behaviour, so I wasn't disappointed when it didn't work.

Could we add support so adding the `!` suffix DTRT? (Or having a `"semanticCommitBreakingChange": true` or similar config option? 🤷‍♂️

**Reply by rarkins** [reply]
Yes, a community PR would be welcome to add ! Support

**Comment by skycaptain** [comment, marked as answer]
Puzzled on why I hadn't realized this earlier, but leaving `"semanticCommitType": "feat"` and defining a commit body works—for example, `"commitBody": "BREAKING CHANGE: major update of {{packageName}}"`. This keeps the commit message title lowercase and allows to trigger breaking changes with the commit. For others reading this, note that some presets also use `commitBody` (e.g., to sign the commit), so please consult the [documentation](https://docs.renovatebot.com/configuration-options/#commitbody) first.
